src/main/scala/chisel3/util/BitPat.scala

```diff
@@ -161,14 +273,83 @@ sealed class BitPat(val value: BigInt, val mask: BigInt, width: Int) extends Sou
     new BitPat((value << that.getWidth) + that.value, (mask << that.getWidth) + that.mask, this.width + that.getWidth)
   }
 
-  /** Generate raw string of a BitPat. */
-  def rawString: String = Seq.tabulate(width) { i =>
+  /** Check whether this [[BitPat]] overlap with that [[BitPat]], i.e. !(intersect.isEmpty)
+    *
+    * @param that [[BitPat]] to be checked.
+    * @return true if this and that [[BitPat]] have overlap.
+    */
+  def overlap(that: BitPat): Boolean = ((mask & that.mask) & (value ^ that.value)) == 0
+
+  /** Check whether this [[BitSet]] covers that (i.e. forall b matches that, b also matches this)
+    *
+    * @param that [[BitPat]] to be covered
+    * @return true if this [[BitSet]] can cover that [[BitSet]]
+    */
+  def cover(that: BitPat): Boolean = (mask & (~that.mask | (value ^ that.value))) == 0
+
+  /** Intersect `this` and `that` [[BitPat]].
+    *
+    * @param that [[BitPat]] to be intersected.
+    * @return a [[BitSet]] containing all elements of `this` that also belong to `that`.
+    */
+  def intersect(that: BitPat): BitSet = {
+    if (!overlap(that)) {
+      BitSet.empty
+    } else {
+      new BitPat(this.value | that.value, this.mask | that.mask, this.width.max(that.width))
+    }
+  }
+
+  /** Subtract a [[BitPat]] from this.
+    *
+    * @param that subtrahend [[BitPat]].
+    * @return a [[BitSet]] containing elements of `this` which are not the elements of `that`.
+    */
+  def subtract(that: BitPat): BitSet = {
+    require(width == that.width)
+    def enumerateBits(mask: BigInt): Seq[BigInt] = {
+      if (mask == 0) {
+        Nil
+      } else {
+        // bits comes after the first '1' in a number are inverted in its two's complement.
+        // therefore bit is always the first '1' in x (counting from least significant bit).
+        val bit = mask & (-mask)
+        bit +: enumerateBits(mask & ~bit)
+      }
+    }
+
+    val intersection = intersect(that)
+    val omask = this.mask
+    if (intersection.isEmpty) {
+      this
+    } else {
+      new BitSet {
+        val terms =
+          intersection.terms.flatMap { remove =>
+            enumerateBits(~omask & remove.mask).map { bit =>
+              // Only care about higher than current bit in remove
+              val nmask = (omask | ~(bit - 1)) & remove.mask
+              val nvalue = (remove.value ^ bit) & nmask
+              val nwidth = remove.width
+              new BitPat(nvalue, nmask, nwidth)
+            }
+          }
+      }
+    }
+  }
+
+  override def isEmpty: Boolean = false
```

Comment:
If this is always false, and terms are always `BitPat`, can `isEmpty` on trait `BitSet` ever be true?

Reply:
Oh. I think here is a bug. Should be mask = 0 here! My bad I’ll add a test to this.

Reply:
`isEmpty` for `BitPat` always returns `false` since `BitPat` cannot represent a semantic that no bit vector should belong can be covered by this `BitPat`. a `mask = 0` represents all bits are `?` which means all bit vector can be covered by this, while `mask = -1`(all bits are 1) represents only one point are covered.
Thus `BitPat` `isEmpty` is always false.

Reply:
In `BitSet` API:
```
def isEmpty: Boolean = terms.forall(_.isEmpty)
```
this means only empty set or a set which terms are all empty set are empty.

Reply:
For detail test, see:
https://github.com/chipsalliance/chisel3/blob/34f313e9d8c30171bc36c7a1f6993fb95efd9dd4/src/test/scala/chiselTests/util/BitSetSpec.scala#L20-L40

Reply:
Okay cool. This makes sense. That being said it'd be nice to see a couple of sanity checks like that `BitSet.empty.isEmpty` is true and that `BitSet("1100").intersect(BitSet("0011")).isEmpty` is also true.

Reply:
Ah didn't see the examples when I responded. Okay good, thanks!